sdk/core/azure-core-http-httpurlconnection/src/main/java/com/azure/core/http/httpurlconnection/HttpUrlConnectionAsyncClient.java

```diff
@@ -0,0 +1,352 @@
+            }
+            assert connection != null;
+            if (connectionTimeout != -1) {
+                connection.setConnectTimeout((int) connectionTimeout);
+            }
+            if (readTimeout != -1) {
+                connection.setReadTimeout((int) readTimeout);
+            }
+            try {
+                connection.setRequestMethod(httpRequest.getHttpMethod().toString());
+            } catch (ProtocolException e) {
+                throw LOGGER.logExceptionAsError(new RuntimeException(e));
+            }
+            for (HttpHeader header : httpRequest.getHeaders()) {
+                for (String value : header.getValues()) {
+                    connection.addRequestProperty(header.getName(), value);
+                }
+            }
+            return connection;
+        } catch (IOException e) {
+            throw LOGGER.logExceptionAsError(new RuntimeException(e));
+        }
+    }
+
+    /**
+     * Asynchronously sends the content of an HttpRequest via an HttpUrlConnection instance.
+     *
+     * @param httpRequest The HTTP Request being sent
+     * @param progressReporter A reporter for the progress of the request
+     * @param connection The HttpURLConnection that is being sent to
+     * @return A Mono that represents the completion of the request sending process
+     */
+    private Mono<Void> sendBodyAsync(HttpRequest httpRequest, ProgressReporter progressReporter, HttpURLConnection connection) {
+        Mono<Void> requestSendMono = Mono.empty();
+
+        switch (httpRequest.getHttpMethod()) {
+            case POST:
+            case PUT:
+            case DELETE: {
+                connection.setDoOutput(true);
+
+                Flux<BinaryData> requestBody;
+                BinaryData binaryBodyData = httpRequest.getBodyAsBinaryData();
+
+                if (binaryBodyData == null) {
+                    requestBody = Flux.just(BinaryData.fromByteBuffer(ByteBuffer.wrap(new byte[0])));
+                } else {
+                    requestBody = Flux.just(binaryBodyData);
+                }
+
+                return requestBody
+                    .flatMap(body -> {
+                        if (progressReporter != null) {
+                            progressReporter.reportProgress(body.toBytes().length);
+                        }
+
+                        return Mono.fromCallable(() -> {
+                            try (DataOutputStream os = new DataOutputStream(new BufferedOutputStream(connection.getOutputStream()))) {
+                                byte[] bytes = body.toBytes();
+                                os.write(bytes);
+                                os.flush();
+                                return Mono.just(body);
+                            } catch (IOException e) {
+                                return FluxUtil.monoError(LOGGER, new RuntimeException(e));
+                            }
+                        }).timeout(writeTimeout);
```

Comment:
A few larger items going on here.

All these operations will eagerly convert the request body into a single `byte[]`, this has both performance and runtime behavior problems. Some services such as Storage allow for requests larger than 2GB which will result in an `OutOfMemoryError` here, and even if they aren't greater than 2GB it could result in a lot of memory being allocated. For this reason, I recommend consuming the request body as an `InputStream`, `BinaryData.toStream()`, with a buffer the write the request.

```java
try (DataOutputStream os = new DataOutputStream(new BufferedOutputStream(connection.getOutputStream()))) {
    InputStream bodyStream = body.toStream();
    byte[] buffer = new byte[8192]; // 8KB is a common default, this can be investigated for better options later.
}
```

With this change progress reporting should be moved inside the request body send so it doesn't turn the whole request into a `byte[]` too.

```java
try (DataOutputStream os = new DataOutputStream(new BufferedOutputStream(connection.getOutputStream()))) {
    InputStream bodyStream = body.toStream();
    byte[] buffer = new byte[8192]; // 8KB is a common default, this can be investigated for better options later.
    int b;
    while ((b = bodyStream.read(buffer) != -1) {
        if (progressReporter != null) {
            progressReporter.reportProgress(b);
        }
     }
}
```

After this there is one last issue with the write timeout. The write timeout is meant to track if any data is being written to the network during the timeout period. The way `timeout` is used it requires the entire request body to be written within the timeout period, which can cause issues with larger request bodies and/or slow networks. This can be changed so that the timeout happens if there isn't a read request of the body steam within a certain time.

```suggestion
byte[] buffer = new byte[8192]; // 8KB is a common default, this can be investigated for better options later.
InputStream bodyStream = binaryDataBody.toStream();
return Flux.using(() -> new DataOutputStream(new BufferedOutputStream(connection.getOutputStream())),
    os -> Flux.generate(() -> os, (network, sink) -> {
        try {
            int read = bodyStream.read(buffer);
        
            // Consuming the request body is complete, signal completion.
            if (read == -1) {
                sink.complete();
                return;
            }
        
            if (progressReporter != null) {
                progressReporter.reportProgress(read);
            }
        
            network.write(buffer);
            network.flush();
            
            sink.next(1); // Dummy value to propagate to trigger timeout to reset on data being sent.
        catch (IOException ex) {
            // Error during sending request, terminate stream.
            sink.error(ex);
        }
    }).timeout(writeTimeout), os -> os.close());
```